File: apps/api/app/schemas/community.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class LinkPreviewOut(BaseModel):
    url: str
    title: str | None
    description: str | None
    image_url: str | None
    site_name: str | None


class CreatePostRequest(BaseModel):
    kind: str  # "question" | "poll"
    title: str = Field(min_length=1, max_length=200)
    body: str | None = Field(default=None, max_length=4000)
    is_anonymous: bool = False
    options: list[str] = Field(default_factory=list, max_length=6)
    attachments: list[str] = Field(default_factory=list, max_length=4)


class VoteRequest(BaseModel):
    value: int  # +1 or -1


class PollVoteRequest(BaseModel):
    option_id: uuid.UUID


class CreateCommentRequest(BaseModel):
    body: str = Field(min_length=1, max_length=2000)
    is_anonymous: bool = False
    attachments: list[str] = Field(default_factory=list, max_length=2)


class HideRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=300)


class PollOptionOut(BaseModel):
    id: uuid.UUID
    label: str
    position: int
    vote_count: int


class CommunityPostRow(BaseModel):
    id: uuid.UUID
    kind: str
    title: str
    excerpt: str | None
    author_display: str
    is_anonymous: bool
    created_at: datetime
    is_hidden: bool
    comment_count: int
    score: int | None  # question posts only
    my_vote: int | None  # question posts only — the caller's own vote, if any
    options: list[PollOptionOut] | None  # poll posts only
    my_option_id: uuid.UUID | None  # poll posts only
    link: LinkPreviewOut | None
    attachments: list[str]


class CommunityCommentRow(BaseModel):
    id: uuid.UUID
    author_display: str
    is_anonymous: bool
    body: str
    attachments: list[str]
    created_at: datetime
    is_hidden: bool
    score: int
    my_vote: int | None


class CommunityPostDetail(CommunityPostRow):
    body: str | None
    comments: list[CommunityCommentRow]

File: apps/web/src/lib/api/community.ts
import { apiFetch, apiUpload } from "./core";

export type CommunityPostKind = "question" | "poll";

export type LinkPreview = {
  url: string;
  title: string | null;
  description: string | null;
  image_url: string | null;
  site_name: string | null;
};

export type PollOption = { id: string; label: string; position: number; vote_count: number };

export type CommunityPostSummary = {
  id: string;
  kind: CommunityPostKind;
  title: string;
  excerpt: string | null;
  author_display: string;
  is_anonymous: boolean;
  created_at: string;
  edited_at: string | null;
  is_hidden: boolean;
  is_mine: boolean;
  comment_count: number;
  score: number | null;
  my_vote: number | null;
  options: PollOption[] | null;
  my_option_id: string | null;
  link: LinkPreview | null;
  attachments: string[];
};

export type CommunityComment = {
  id: string;
  author_display: string;
  is_anonymous: boolean;
  body: string;
  attachments: string[];
  created_at: string;
  edited_at: string | null;
  is_hidden: boolean;
  is_mine: boolean;
  score: number;
  my_vote: number | null;
};

export type CommunityPostDetail = CommunityPostSummary & {
  body: string | null;
  comments: CommunityComment[];
};

export const communityApi = {
  list: (kind?: CommunityPostKind) => apiFetch<CommunityPostSummary[]>(`/community/posts${kind ? `?kind=${kind}` : ""}`),
  get: (id: string) => apiFetch<CommunityPostDetail>(`/community/posts/${id}`),
  create: (payload: {
    kind: CommunityPostKind;
    title: string;
    body: string | null;
    is_anonymous: boolean;
    options: string[];
    attachments: string[];
  }) => apiFetch<CommunityPostSummary>("/community/posts", { method: "POST", body: JSON.stringify(payload) }),
  update: (
    id: string,
    payload: { title: string; body: string | null; is_anonymous: boolean; attachments: string[] }
  ) => apiFetch<CommunityPostSummary>(`/community/posts/${id}`, { method: "PATCH", body: JSON.stringify(payload) }),
  remove: (id: string) => apiFetch<void>(`/community/posts/${id}`, { method: "DELETE" }),
  vote: (id: string, value: 1 | -1) =>
    apiFetch<CommunityPostSummary>(`/community/posts/${id}/vote`, { method: "POST", body: JSON.stringify({ value }) }),
  pollVote: (id: string, optionId: string) =>
    apiFetch<CommunityPostSummary>(`/community/posts/${id}/poll-vote`, {
      method: "POST",
      body: JSON.stringify({ option_id: optionId }),
    }),
  addComment: (id: string, body: string, isAnonymous: boolean, attachments: string[]) =>
    apiFetch<CommunityPostDetail>(`/community/posts/${id}/comments`, {
      method: "POST",
      body: JSON.stringify({ body, is_anonymous: isAnonymous, attachments }),
    }),
  updateComment: (id: string, payload: { body: string; is_anonymous: boolean; attachments: string[] }) =>
    apiFetch<CommunityComment>(`/community/comments/${id}`, { method: "PATCH", body: JSON.stringify(payload) }),
  voteComment: (id: string, value: 1 | -1) =>
    apiFetch<CommunityComment>(`/community/comments/${id}/vote`, { method: "POST", body: JSON.stringify({ value }) }),
  uploadAttachment: (file: File) => apiUpload<{ url: string }>("/community/uploads", file),
  hidePost: (id: string, reason?: string) =>
    apiFetch<void>(`/community/posts/${id}/hide`, { method: "POST", body: JSON.stringify({ reason: reason ?? null }) }),
  unhidePost: (id: string) => apiFetch<void>(`/community/posts/${id}/unhide`, { method: "POST" }),
  hideComment: (id: string, reason?: string) =>
    apiFetch<void>(`/community/comments/${id}/hide`, { method: "POST", body: JSON.stringify({ reason: reason ?? null }) }),
  unhideComment: (id: string) => apiFetch<void>(`/community/comments/${id}/unhide`, { method: "POST" }),
};
